src/rgw/rgw_rest_s3.cc

```diff
@@ -4393,30 +4385,122 @@ RGWAuthApplier::aplptr_t RGWLDAPAuthEngine::authenticate() const
   RGWUserInfo user_info;
   user_info.user_id = base64_token.id;
   if (rgw_get_user_info_by_uid(store, user_info.user_id, user_info) >= 0) {
-    if (user_info.type != TYPE_LDAP) {
+    /*if (user_info.type != TYPE_LDAP) {
       ldout(cct, 10) << "ERROR: User id of type: " << user_info.type << " is already present" << dendl;
       return nullptr;
-    }
+    }*/
   }
 
   if (ldh->auth(base64_token.id, base64_token.key) != 0) {
-    return nullptr;
+    throw -EACCES;
   }
 
   return apl_factory->create_apl_remote(cct, get_acl_strategy(), get_creds_info(base64_token));
 }
 
-std::string RGWLDAPTokenExtractor::get_token() const
+void RGWS3V2Extractor::get_tokens(std::string& access_key_id,
+                                  std::string& signature,
+                                  std::string& expires,
+                                  bool& qsr) const
 {
-  string token = "";
+  qsr = false;
   if (!s->http_auth || !(*s->http_auth)) {
-    token = s->info.args.get("AWSAccessKeyId");
+    access_key_id = s->info.args.get("AWSAccessKeyId");
+    signature = s->info.args.get("Signature");
+    expires = s->info.args.get("Expires");
+    qsr = true;
   } else {
     string auth_str(s->http_auth + 4);
     int pos = auth_str.rfind(':');
-    if (pos >=0 ) {
-      token = auth_str.substr(0, pos);
+    if (pos >= 0) {
+      access_key_id = auth_str.substr(0, pos);
+      signature = auth_str.substr(pos + 1);
+    }
+  }
+}
+
+bool RGWS3V2LocalAuthEngine::is_applicable() const noexcept
+{
+  if (! RGWS3V2AuthEngine::is_applicable()) {
+    return false;
+  }
+  if (! s->cct->_conf->rgw_s3_auth_use_rados) {
+    return false;
+  }
+
+  return true;
+}
+
+RGWAuthApplier::aplptr_t RGWS3V2LocalAuthEngine::authenticate() const
+{
+  if (access_key_id.empty() || signature.empty()) {
+    ldout(s->cct, 5) << "access_key_id or signature is empty" << dendl;
+    throw -EINVAL;
+  }
+
+  time_t now;
+  time(&now);
+  if (! expires.empty()) {
+    time_t exp = atoll(expires.c_str());
+    if (now >= exp) {
+      throw -EPERM;
     }
   }
-  return token;
+  /* get the user info */
+  string access_key = access_key_id;
+  if (rgw_get_user_info_by_access_key(store, access_key, *(s->user)) < 0) {
+      ldout(s->cct, 5) << "error reading user info, uid=" << access_key_id
+              << " can't authenticate" << dendl;
+      throw -EPERM;
+  } /*else {
+    if (s->user->type != TYPE_RGW) {
+      ldout(cct, 10) << "ERROR: User id of type: " << s->user->type << " is present" << dendl;
+      throw -EPERM;
+    }
+  }*/
+
```

Comment:
Same comment as above, this will be enabled when the users are migrated to the new type.